File: Organize.py
import os
import shutil
import re

path = input("Enter the path : ")
files = [f for f in os.listdir(path) if os.path.isfile(os.path.join(path, f))]

def get_season(filename:str):
    pattern_regex = r"([Ss](\d+))"
    a=re.findall(pattern_regex, filename)
    if len(a) > 0:
        return "S"+str("{:02d}".format(int(a[0][1])))
    else:
        return None

def get_resolution(filename:str):
    pattern_regex = r"((480|1080|720|2160)[Pp])"
    a=re.findall(pattern_regex, filename)
    if len(a) > 0:
        return a[0][0].lower()
    else:
        return None

def get_encode(filename:str):
    text = ""
    if re.search("[Xx]265",filename) != None:
        text+="X265"
    if re.search(r"(10.?[Bb][Ii][Tt])",filename) or re.search(r"[Hh][Ee][Vv][Cc]",filename) != None:
        text+=" 10Bit" if text != "" else "10Bit"
    if text =="":
        return None
    return text

def get_type(filename:str):
    if re.search(r"[Dd][Uu][Bb][Bb]?[Ll]?[Ee][Dd]?",filename) != None:
        return "Dubbled"
    elif re.search(r"[Ss][Oo|Uu][Ff|Bb][Tt]?[Bb]?[Ee]?[Dd]?",filename) != None:
        return "SoftSub"
    else:
        return None

def get_file_path(filename:str):
    newpath = path 
    season = get_season(filename)
    if season != None:
        newpath += f"/{season}"
    type = get_type(filename)
    if type != None:
        newpath += f"/{type}"
    resolution = get_resolution(filename)
    if resolution != None:
        newpath += f"/{resolution}"
    encodes= get_encode(filename)
    if encodes != None:
        newpath += f" {encodes}"

    return newpath
    
def create_folder_season(files:list):
    fol =[]
    for i in files:
        season = get_season(i)
        if season != None:
            if season not in fol:
                fol.append(season)
                new = path+f"/{season}"
                if os.path.exists(new) != True:
                    os.mkdir(new)

def create_folder_type(files:list):
    for i in files:
        type = get_type(i)
        if type != None:
            season = get_season(i)
            if season != None:
                    new = path+f"/{season}/{type}"
                    if os.path.exists(new) != True:
                        os.mkdir(new)

def create_folder_resolution(files:list):
    newpath = ""
    for i in files:
        resolution = get_resolution(i)
        encodes= get_encode(i)
        if resolution != None:
            newpath = resolution

        if encodes != None:
            newpath += f" {encodes}"

        season = get_season(i)
        if season != None:
            gtype= get_type(i)
            if gtype != None:
                season += f"/{gtype}"   
            new = path+f"/{season}/{newpath}"
            if os.path.exists(new) != True:
                os.mkdir(new)

def move_files(files:list):
    for i in files:
        newpath = get_file_path(i)
        if os.path.exists(newpath) == True:
            shutil.move(path+f"/{i}",newpath+f"/{i}")

if files != []:
    create_folder_season(files)
    create_folder_type(files)
    create_folder_resolution(files)
    move_files(files)
else:
    print("No files found")

File: GO/Organize.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	var path string
	fmt.Print("Enter the path : ")
	fmt.Scanln(&path)
	files := GetSerialFiles(path)

	CreateFolders(files, path)
	MoveFiles(files, path)
}

func GetSerialFiles(dir string) []string {
	var files []string
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			fmt.Println(err)
			return nil
		}
		if info.IsDir() == false {
			filename := filepath.Base(path)
			if GetSeason(filename) != "" {
				if GetResolution(filename) != "" {
					files = append(files, filename)
				}
			}
		}
		return nil
	})
	if err != nil {
		fmt.Println(err)
	}
	return files
}

func GetSeason(filename string) string {
	getSeason := regexp.MustCompile(`([Ss](\d+))`)
	season := getSeason.FindStringSubmatch(filename)
	if len(season) > 0 {
		st, _ := strconv.Atoi(season[2])
		s := fmt.Sprintf("%02d", st)
		return "S" + s
	}
	return ""
}

func GetResolution(filename string) string {
	getResolution := regexp.MustCompile(`((480|1080|720|2160)[Pp])`)
	resolution := getResolution.FindStringSubmatch(filename)
	if len(resolution) > 0 {
		resolution[0] = strings.TrimSpace(resolution[0])
		return strings.ToLower(resolution[0])
	}
	return ""
}

func GetEncode(filename string) string {
	var text = ""
	getEncode1 := regexp.MustCompile(`[Xx]265`)
	getEncode2 := regexp.MustCompile(`(10.?[Bb][Ii][Tt])`)
	getEncode3 := regexp.MustCompile(`[Hh][Ee][Vv][Cc]`)
	x256 := getEncode1.FindStringSubmatch(filename)
	bit10 := getEncode2.FindStringSubmatch(filename)
	hevc := getEncode3.FindStringSubmatchIndex(filename)

	if len(x256) > 0 {
		text += x256[0]
	}
	if len(bit10) > 0 || len(hevc) > 0 {
		if text == "" {
			text += "10Bit"
		} else {
			text += " 10Bit"
		}
	}
	return text
}

func GetType(filename string) string {
	isDubbled := regexp.MustCompile(`[Dd][Uu][Bb][Bb]?[Ll]?[Ee][Dd]?`)
	isSoftSub := regexp.MustCompile(`[Ss][Oo|Uu][Ff|Bb][Tt]?[Bb]?[Ee]?[Dd]?`)
	if isDubbled.MatchString(filename) {
		return "Dubbled"
	}
	if isSoftSub.MatchString(filename) {
		return "SoftSub"
	}
	return ""
}

func GetNewPath(filename string, path string) string {
	newpath := path
	if season := GetSeason(filename); season != "" {
		newpath += "/" + season
	}
	if filetype := GetType(filename); filetype != "" {
		newpath += "/" + filetype
	}
	if res := GetResolution(filename); res != "" {
		newpath += "/" + res
	}
	if en := GetEncode(filename); en != "" {
		newpath += " " + en
	}

	return newpath
}

func CreateFolders(files []string, path string) {

	for _, file := range files {
		newpath := GetNewPath(file, path)
		os.MkdirAll(newpath, os.ModeDir)
	}
}

func MoveFiles(files []string, path string) bool {
	for _, file := range files {
		oldpath := path + "/" + file
		newpath := GetNewPath(file, path) + "/" + file
		err := os.Rename(oldpath, newpath)
		if err != nil {
			return false
		}
	}
	return true
}
